### ISSUE TITLE
Timeout during HTTP2 upgrade 

### ISSUE THREAD

**Issue by dimaqq** [pgjones/hypercorn#51]
I've observed that when an endpoint delays before responding, like so:

```py
import asyncio

from fastapi import FastAPI


app = FastAPI()


@app.get("/")
async def foobar():
    await asyncio.sleep(9)
    return {"foo": "bar"}
```

Details: I'm running `hypercorn main:app` without certs, so I'm testing clear-text `h2c`

When I talk to this endpoint with `curl` with `--http2-prior-knowldge`, it all works.

But when I ask `curl` to do the HTTP/2 upgrade, `--http2` flag, the server `hypercorn` appears to close the connection after 5 seconds.

MRE: https://github.com/dimaqq/repro-task-destroyed/tree/repro-http-upgrade

Here's `strace` for `curl`:

```c
sendto(5, "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n", 24, MSG_NOSIGNAL, NULL, 0) = 24
sendto(5, "\0\0\22\4\0\0\0\0\0\0\3\0\0\0d\0\4\2\0\0\0\0\2\0\0\0\0", 27, MSG_NOSIGNAL, NULL, 0) = 27
sendto(5, "\0\0\4\10\0\0\0\0\0\1\377\0\1", 13, MSG_NOSIGNAL, NULL, 0) = 13
recvfrom(5, "\0\0*\4\0\0\0\0\0\0\1\0\0\20\0\0\2\0\0\0\0\0\4\0\0\377\377\0\5\0\0@"..., 32768, 0, NULL, NULL) = 51
sendto(5, "\0\0\0\4\1\0\0\0\0", 9, MSG_NOSIGNAL, NULL, 0) = 9
poll([{fd=5, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, NULL, 8) = 0
poll([{fd=5, events=POLLIN}, {fd=3, events=POLLIN}], 2, 198) = 1 ([{fd=5, revents=POLLIN}])
rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, 8) = 0
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, NULL, 8) = 0
poll([{fd=5, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 1 ([{fd=5, revents=POLLIN|POLLRDNORM}])
recvfrom(5, "\0\0\0\4\1\0\0\0\0", 32768, 0, NULL, NULL) = 9
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, NULL, 8) = 0
poll([{fd=5, events=POLLIN}, {fd=3, events=POLLIN}], 2, 198) = 0 (Timeout)
rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, 8) = 0
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, NULL, 8) = 0
poll([{fd=5, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, NULL, 8) = 0
poll([{fd=5, events=POLLIN}, {fd=3, events=POLLIN}], 2, 1000) = 0 (Timeout)
rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, 8) = 0
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, NULL, 8) = 0
poll([{fd=5, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, NULL, 8) = 0
poll([{fd=5, events=POLLIN}, {fd=3, events=POLLIN}], 2, 1000) = 0 (Timeout)
rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, 8) = 0
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, NULL, 8) = 0
poll([{fd=5, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, NULL, 8) = 0
poll([{fd=5, events=POLLIN}, {fd=3, events=POLLIN}], 2, 1000) = 0 (Timeout)
rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, 8) = 0
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, NULL, 8) = 0
poll([{fd=5, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, NULL, 8) = 0
poll([{fd=5, events=POLLIN}, {fd=3, events=POLLIN}], 2, 1000) = 0 (Timeout)
rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, 8) = 0
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, NULL, 8) = 0
poll([{fd=5, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, NULL, 8) = 0
poll([{fd=5, events=POLLIN}, {fd=3, events=POLLIN}], 2, 1000) = 1 ([{fd=5, revents=POLLIN}])
rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, 8) = 0
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f96a252b3c0}, NULL, 8) = 0
poll([{fd=5, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 1 ([{fd=5, revents=POLLIN|POLLRDNORM}])
recvfrom(5, "", 32768, 0, NULL, NULL)   = 0
```

It appears to show that `curl` has sent the HTTP/2 connect preface, and is waiting for response.
If the endpoint is "slow/delayed", it seems that `hypercorn` closes the connection (`POLLRDNORM`, `recvfrom() == 0`).

If the fastapi endpoint were to return the response within 5 seconds, the response would be delivered correctly to `curl`.

If the response is streaming, the first 5 seconds of the response are delivered, and then the connection appears to be terminated by `hypercorn`. The endpoint receives `CancelledError`.

**Comment by dimaqq** [comment]
Looks like it's the keep-alive timeout, I can change the time when hyprecorn drops the connection with `hypercorn --keep-alive N`.

Given that HTTP/2 has own ping semantics, I guess hypercorn should disable HTTP/1 keep-alive timer when it receives (or accepts) the HTTP/2 upgrade message.

**Comment by pgjones** [comment]
Thanks, fixed in da9262d75feff7ec068d5d4ef25adc84224fcb7b
